heartbeat/ocf.py

```diff
@@ -135,3 +135,256 @@ def emit(self, record):
+				return s + ' {}="{}"'.format(name, var)
+			return s
+		ret = '<action name="{}"'.format(self.name)
+		ret = opt(ret, "timeout", self.timeout)
+		ret = opt(ret, "interval", self.interval)
+		ret = opt(ret, "depth", self.depth)
+		ret += " />\n"
+		return ret
+
+
+class Metadata(object):
+	"""
+	Metadata XML generator helper.
+	"""
+
+	def __init__(self, name, shortdesc, longdesc):
+		self.name = name
+		self.shortdesc = shortdesc
+		self.longdesc = longdesc
+		self.parameters = []
+		self.actions = []
+
+	def parameter(self, name, shortdesc="", longdesc="", content_type="string", unique=False, required=False, default=None):
+		self.parameters.append(Parameter(name=name,
+										 shortdesc=shortdesc,
+										 longdesc=longdesc,
+										 content_type=content_type,
+										 unique=unique,
+										 required=required,
+										 default=default))
+		return self
+
+	def action(self, name, timeout=None, interval=None, depth=None):
+		self.actions.append(Action(name=name,
+								   timeout=timeout,
+								   interval=interval,
+								   depth=depth))
+		return self
+
+	def __str__(self):
+		return """<?xml version="1.0"?>
+<!DOCTYPE resource-agent SYSTEM "ra-api-1.dtd">
+<resource-agent name="{name}">
+<version>1.0</version>
+<longdesc lang="en">
+{longdesc}
+</longdesc>
+<shortdesc lang="en">{shortdesc}</shortdesc>
+
+<parameters>
+{parameters}
+</parameters>
+
+<actions>
+{actions}
+</actions>
+
+</resource-agent>
+""".format(name=self.name,
+		   longdesc=self.longdesc,
+		   shortdesc=self.shortdesc,
+		   parameters="".join(str(p) for p in self.parameters),
+		   actions="".join(str(a) for a in self.actions))
+
+
+def run(metadata, handlers):
+	"""
+	Main loop implementation for resource agents.
+	Does not return.
+
+	Arguments:
+
+	metadata: Metadata structure generated by ocf.Metadata
+
+	handlers: Dict of action name to handler function.
+
+	Handler functions can take parameters as arguments,
+	the run loop will read parameter values from the
+	environment and pass to the handler.
+	"""
+	import inspect
+
+	def check_required_params():
+		for p in metadata.parameters:
+			if p.required and parameter(p.name) is None:
+				ocf_exit_reason("{}: Required parameter not set".format(p.name))
+				sys.exit(OCF_ERR_CONFIGURED)
+
+	def call_handler(name, func):
+		if hasattr(inspect, 'signature'):
+			params = inspect.signature(func).parameters.keys()
+		else:
+			params = inspect.getargspec(func).args
+		arglist = [parameter(p, p.default) for p in params]
+		rc = func(*arglist)
+		if rc is None:
+			rc = OCF_SUCCESS
+		return rc
+
+	if len(sys.argv) == 2:
+		action = sys.argv[1]
+	else:
+		action = os.environ.get("__OCF_ACTION")
+	if action is None:
+		ocf_exit_reason("No action argument set")
+		sys.exit(OCF_ERR_UNIMPLEMENTED)
+	if action in ('meta-data', 'usage', 'methods'):
+		sys.stdout.write(str(metadata) + "\n")
+		sys.exit(OCF_SUCCESS)
+
+	check_required_params()
+	for handler, func in handlers.items():
```

Comment:
I think it'd be smarter to do "if action in handlers:" instead of the for-loop.

Reply:
yeah, I'll change it.